File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  const noDuplicates = new Set()
  for (const val of arr) {
    if (noDuplicates.has(val)) {
      return val
    }
    noDuplicates.add(val)
  }
  return -1
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));

  console.log("");

  console.log("Expecting: 7");
  console.log("=>", findFirstDuplicate([8, 7, 7, 9]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([4, 3, 2, 1]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// make empty set to put in values - iterate through array if unique value put into the set, if value matches something in the set return that value

// And a written explanation of your solution
// 

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
require 'set'
def find_first_duplicate(arr)
  noDuplicates = Set.new 

  arr.each do |val|
    return val if noDuplicates.include?(val)
    noDuplicates.add(val)
  end
  -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([8, 7, 7, 9]);

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([4, 3, 2, 1]);
  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution
